nucypher/policy/policies.py

```diff
@@ -165,30 +165,100 @@ def payload(self):
+        self.released = False
+        self.when_complete = Deferred()  # TODO: Allow cancelling via KB Interrupt or some other way?
+
+        if note is None:
+            tp_name=f"{f} to {len(nodes)} nodes"
+        else:
+            tp_name = f"{note}: {f} to {len(nodes)} nodes"
 
-        if self.finishedCount == self._block_until_this_many_are_complete:
-            self._policy_locking_queue.put(None)  # TODO: It'd be rad to return a list of nodes here who were contacted, but it's non-trivial.
-        return result
+        self._threadpool = ThreadPool(minthreads=threadpool_size, maxthreads=threadpool_size, name=tp_name)
 
     def block_for_a_little_while(self):
         """
         https://www.youtube.com/watch?v=OkSLswPSq2o
         """
-        _ = self._policy_locking_queue.get()  # Interesting opportuntiy to pass some data, like the list of contacted nodes above.
-        self.log.debug(f"{self.finishedCount} nodes were contacted while blocking for a little while.")
-        return
+        if len(self.completed) < self._block_until_this_many_are_complete:
+            _ = self._partial_queue.get()  # Interesting opportuntiy to pass some data, like the list of contacted nodes above.
+            self.log.debug(f"{len(self.completed)} nodes were contacted while blocking for a little while.")
+
+    def block_until_complete(self):
+        if self.total_disposed() < len(self.nodes):
+            _ = self._completion_queue.get()  # Interesting opportuntiy to pass some data, like the list of contacted nodes above.
+
+    def _handle_success(self, response, node):
+        if response.status_code == 202:
+            self.completed[node] = response
+        else:
+            assert False  # TODO: What happens if this is a 300 or 400 level response?
+
+        if len(self.completed) == self._block_until_this_many_are_complete:
+            print(f"++++++++++++++BLOCKED FOR A LITTLE WHILE, completed {len(self.completed)} nodes")
+            self._partial_queue.put(self.completed)
+            self.released = True
+        self._consider_finalizing()
+        return response
+
+    def _handle_error(self, failure, node):
+        self.failed[node] = failure  # TODO: Add a failfast mode?
+        self._consider_finalizing()
+        print(f"{datetime.datetime.now()}: !!!!!!!!FAILED {node}")
+
+    def total_disposed(self):
+        return len(self.completed) + len(self.failed)
+
+    def _consider_finalizing(self):
+        if self.total_disposed() >= len(self.nodes):
+            self._completion_queue.put(self.completed)
+            self.when_complete.callback(self.completed)
+            self._threadpool.stop()
+
+    def _engage_node(self, node):
+        maybe_coro = self.f(node, network_middleware=self.network_middleware, *self.args, **self.kwargs)
+
+        d = ensureDeferred(maybe_coro)
+        d.addCallback(self._handle_success, node)
+        d.addErrback(self._handle_error, node)
+        return d
+
+    def start(self):
```

Comment:
Perhaps an associated `stop()` function could be provided that unblocks everything , stops the threads and threadpool could be good for forcibly stopping the mutex in extreme cases...particularly in failed tests?

Reply:
:+1: 

Care to open an issue?

Reply:
Filed #2189